odoo/models.py

```diff
@@ -4445,11 +4451,89 @@ def unlink(self):
+                sub_ids_json_text = tuple(str(id_) for id_ in sub_ids)
+
+                if self._context.get(MODULE_UNINSTALL_FLAG):
+                    self.env.execute_query(SQL(
+                        """
+                        DELETE FROM ir_default
+                        WHERE field_id IN %(field_ids)s
+                        AND json_value IN %(ids)s
+                        """,
+                        field_ids=field_ids,
+                        ids=sub_ids_json_text,
+                    ))
+                elif res := self.env.execute_query(SQL(
+                    """
+                    SELECT json_value::int4, field_id
+                    FROM ir_default
+                    WHERE field_id IN %(field_ids)s
+                    AND json_value IN %(ids)s
+                    LIMIT 1
+                    """,
+                    field_ids=field_ids,
+                    ids=sub_ids_json_text,
+                )):
+                    record_id, field_id = res[0]
+                    ir_field = self.env['ir.model.fields'].browse(field_id).sudo()
+                    field = self.env[ir_field.model]._fields[ir_field.name]
+                    record = self.browse(record_id)
+                    raise UserError(_('Unable to delete %(record)s because it is used as the default value of %(field)s', record=record, field=field))
+
+            # on delete set null/restrict for jsonb company dependent many2one
+            for field in self.env.registry.many2one_company_dependents[self._name]:
+                if field.ondelete == 'restrict' and not self._context.get(MODULE_UNINSTALL_FLAG):
+                    if res := self.env.execute_query(SQL(
+                        """
+                        SELECT id, %(field)s
+                        FROM %(table)s
+                        WHERE %(field)s IS NOT NULL
+                        AND %(field)s @? %(jsonpath)s
+                        LIMIT 1
+                        """,
+                        table=SQL.identifier(self.env[field.model_name]._table),
+                        field=SQL.identifier(field.name),
+                        jsonpath=f"$.* ? ({' || '.join(f'@ == {id_}' for id_ in sub_ids)})",
+                    )):
+                        on_restrict_id, field_json = res[0]
+                        to_delete_id = next(iter(id_ for id_ in field_json.values()))
+                        on_restrict_record = self.env[field.model_name].browse(on_restrict_id)
+                        to_delete_record = self.browse(to_delete_id)
+                        raise UserError(_('You cannot delete %(to_delete_record)s, as it is used by %(on_restrict_record)s',
+                                          to_delete_record=to_delete_record, on_restrict_record=on_restrict_record))
+                else:
+                    self.env.execute_query(SQL(
+                        """
+                        UPDATE %(table)s
```

Comment:
You need to invalidate the cache of this field then.

Reply:
`self.env.invalidate_all(flush=False)` will do that at the end of the `unlink`